lib/python/picongpu/picmi/copy_attributes.py

```
@@ -0,0 +1,170 @@
+
+    This class decorator adds a method `get_as_pypicongpu`
+    that copies the content of `self` to a new instance of `to_class`
+    by virtue of the `copy_attributes` function.
+    See the `copy_attributes` docstring for details.
+
+    If given, it runs `preamble(self)` before copying.
+    """
+    if has_attribute(to_class, "get_as_pypicongpu"):
+        message = f"Adding 'get_as_pypicongpu' failed because it already existed on receiving class {to_class}."
+        raise TypeError(message)
+
+    def decorator(cls):
+        def get_as_pypicongpu(self, *args, **kwargs):
+            if preamble:
+                preamble(self, *args, **kwargs)
+            local_conversions = {
+                key: value
+                if isinstance(value, str)
+                # Python binds lambdas by reference, so if we were to use the seemingly obvious implementation,
+                # all lambdas in this dictionary would use the last value of `value`.
+                # That is why we need to force immediate evaluation:
+                else (lambda local_value: lambda self: local_value(self, *args, **kwargs))(value)
+                for key, value in (conversions or {}).items()
+            }
+            return copy_attributes(
+                self,
+                to_class,
+                conversions=local_conversions,
+                remove_prefix=remove_prefix,
+                ignore=ignore,
+                default_converter=lambda self: default_converter(self, *args, **kwargs),
+            )
+
+        cls.get_as_pypicongpu = get_as_pypicongpu
+        return cls
+
+    return decorator
+
+
+def default_converts_to(to_class, conversions=None, preamble=None, remove_prefix="", ignore=tuple()):
+    return converts_to(
+        to_class,
+        conversions=conversions,
+        preamble=preamble
+        or (lambda self, *args, **kwargs: self.check(*args, **kwargs) if has_attribute(self, "check") else None),
```

The current design choice is to run `check`, if it exists, if no preamble is provided. If it is provided, the user is responsible to run `check` if they want it. 
This is fine, but I was wondering if we should always check, separate it from the preamble, and make it opt out.

---

Mmmmhhhh... I'm not so much in favour of this whole `check` business anyways. I'd prefer to have the constructor handle argument normalisation and let it raise an exception if it fails to do so. In that sense, I have no inclination to lean into that further by defining an (implicit) protocol around this.

---

Detecting and running `check` is mostly a concession to the current reality of the code and an attempt to separate concerns. Otherwise I'd have restructured the code even more.

---

the reason why our PICMI implementation has `check()` methods is that otherwise you get 300+ line constructors, which carries with its own baggage.
There are ways to avoid a lot of the checks using pydantic, but that requires a mayor refactor of the code base, something I am planning to do some time in the future, but not right now.